ts-client/src/vault/index.ts

```diff
@@ -0,0 +1,330 @@
+import { getStrategyHandler, getStrategyType, StrategyState } from "./strategy";
+import { IDL, Vault as VaultIdl } from "./idl";
+
+const getVaultState = async (
+    vaultParams: TokenInfo,
+    program: VaultProgram
+): Promise<{ vaultPda: PublicKey, tokenVaultPda: PublicKey, vaultState: VaultState, lpSupply: string }> => {
+    const { vaultPda, tokenVaultPda } = await getVaultPdas(new PublicKey(vaultParams.address), new PublicKey(program.programId));
+    const vaultState = (await program.account.vault.fetchNullable(
+        vaultPda
+    )) as VaultState;
+    const lpSupply = await getLpSupply(program.provider.connection, vaultState.lpMint);
+
+    if (!vaultState) {
+        throw "Cannot get vault state";
+    }
+    return { vaultPda, tokenVaultPda, vaultState, lpSupply };
+}
+
+const getVaultLiquidity = async (connection: Connection, tokenVaultPda: PublicKey): Promise<string | null> => {
+    const vaultLiquidityResponse = await connection.getAccountInfo(tokenVaultPda)
+    if (!vaultLiquidityResponse) return null;
+
+    const vaultLiquidtySerialize = deserializeAccount(vaultLiquidityResponse.data)
+    return vaultLiquidtySerialize?.amount.toString() || null;
+}
+
+export default class VaultImpl implements VaultImplementation {
+    private connection: Connection;
+    private cluster: Cluster = 'mainnet-beta';
+
+    // Vault
+    private tokenInfo: TokenInfo;
+    private program: VaultProgram;
+
+    public vaultPda: PublicKey;
+    public tokenVaultPda: PublicKey;
+    public vaultState: VaultState;
+    public lpSupply: string = '';
+
+    private constructor(program: VaultProgram, vaultDetails: VaultDetails, opt?: { cluster?: Cluster }) {
+        this.connection = program.provider.connection;
+        this.cluster = opt?.cluster ?? 'mainnet-beta';
+
+        this.tokenInfo = vaultDetails.vaultParams;
+        this.program = program;
+        this.vaultPda = vaultDetails.vaultPda;
+        this.tokenVaultPda = vaultDetails.tokenVaultPda;
+        this.vaultState = vaultDetails.vaultState;
+        this.lpSupply = vaultDetails.lpSupply;
+    }
+
+    public static async create(
+        connection: Connection,
+        vaultParams: TokenInfo,
+        opt?: {
+            cluster?: Cluster,
+            programId?: string,
+        }
+    ): Promise<VaultImpl> {
+        const provider = new AnchorProvider(connection, {} as any, AnchorProvider.defaultOptions());
+        const program = new Program<VaultIdl>(
+            IDL as VaultIdl,
+            opt?.programId || PROGRAM_ID,
+            provider
+        );
+
+        const { vaultPda, tokenVaultPda, vaultState, lpSupply } = await getVaultState(vaultParams, program);
+        return new VaultImpl(program, { vaultParams, vaultPda, tokenVaultPda, vaultState, lpSupply }, opt);
+    }
+
+    public async getUserBalance(owner: PublicKey): Promise<string> {
+        const address = await getAssociatedTokenAccount(this.vaultState.lpMint, owner);
```

Comment:
This make sense when it is used directly by the user. However, most of the integrated program will not use the seed that is same as associated token account. Maybe instead of passing owner key, we allow them to pass in any public key, and verify the mint address ?
eg:
https://github.com/mercurial-finance/mercurial-dynamic-amm/blob/main/programs/amm/src/context.rs#L80